{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf055b67",
   "metadata": {},
   "source": [
    "## Run MMseqs2 on the single mutation benchmark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a23a7162",
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "import sys\n",
    "import csv\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "root_dir = Path('/local/sieg/projekte/microminer_evaluation')\n",
    "sys.path.insert(0, str(root_dir.resolve()))\n",
    "\n",
    "import helper\n",
    "from helper.data_operations import read_microminer_csv, merge_results_for_pair_eval\n",
    "from helper import BAD_PDBIDS\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a96c845",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_collection = helper.get_dataset_collection()\n",
    "supported_datasets = [\n",
    "    dataset.name\n",
    "    for dataset in dataset_collection.get_mutation_datasets_with_structure_pairs()\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff9cf517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating known mutations of protherm\n",
      "Evaluating known mutations of platinum\n",
      "Evaluating known mutations of thermomutdb\n",
      "Evaluating known mutations of shanthirabalan\n",
      "collected 123 unique query ids\n"
     ]
    }
   ],
   "source": [
    "query_ids_set = set() \n",
    "for dataset_name in supported_datasets:\n",
    "    dataset = dataset_collection.get_dataset(dataset_name)\n",
    "    print(f\"Evaluating known mutations of {dataset.name}\")\n",
    "    df_ref = dataset.read_single_mutations(pdb_mutant_only=True)\n",
    "    df_ref = df_ref[~df_ref['wild_pdb'].isin(BAD_PDBIDS)]\n",
    "    df_ref = df_ref[~df_ref['mut_pdb'].isin(BAD_PDBIDS)]\n",
    "    query_ids_set.update(df_ref['wild_pdb'].str.lower() + '_' + (df_ref['wild_chain'] if 'wild_chain' in df_ref.columns else 'A'))\n",
    "\n",
    "# only sample in protherm that is not chain 'A' is 2CI2. Just add manually\n",
    "query_ids_set.remove('2ci2_A')\n",
    "query_ids_set.add('2ci2_I')\n",
    "\n",
    "# in platinum the PDB files are customly prepared. We have to adjust the chain identifier to match the seqres sequences\n",
    "query_ids_set.remove('3pca_I')\n",
    "query_ids_set.add('3pca_A')\n",
    "\n",
    "print(f'collected {len(query_ids_set)} unique query ids')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "610b9561",
   "metadata": {},
   "outputs": [],
   "source": [
    "dtypes_dict = {'pdb_wild': str, 'pdb_mutant': str, 'wild_chain': str, 'wild_aa': str, 'seq_num': str, 'mutation_aa': str, 'reason': str}\n",
    "df_problematic = pd.read_csv(root_dir / 'data' / 'problematic_single_mutations.tsv', sep='\\t', header=0, dtype=dtypes_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7de1a811",
   "metadata": {},
   "source": [
    "Download the PDB seqres fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf332d4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORK_DIR = Path(\"/local/sieg/mmseqs_workdir\")\n",
    "QUERY_SET_PATH = WORK_DIR / 'pdb_seqres_query.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f9075077",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-26 10:31:13 URL:https://ftp.wwpdb.org/pub/pdb/derived_data/pdb_seqres.txt.gz [47185569/47185569] -> \"/local/sieg/mmseqs_workdir/pdb_seqres.txt.gz\" [1]\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "\n",
    "WORK_DIR=\"/local/sieg/mmseqs_workdir\"\n",
    "mkdir -p ${WORK_DIR}\n",
    "\n",
    "# download seqres from PDB\n",
    "wget https://ftp.wwpdb.org/pub/pdb/derived_data/pdb_seqres.txt.gz -P ${WORK_DIR} -nv\n",
    "gunzip \"${WORK_DIR}/pdb_seqres.txt.gz\"\n",
    "\n",
    "# # download PDB100 from colabfold\n",
    "# wget https://wwwuser.gwdg.de/~compbiol/colabfold/pdb100_230517.fasta.gz -P ${WORK_DIR} -nv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93179c5e",
   "metadata": {},
   "source": [
    "extract sequences relevant for us"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7fb80d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def extract_sequences(fasta_filename: str, pdb_ids: set[str], out_fasta_filehandle):\n",
    "    assert len(pdb_ids) > 0\n",
    "    id_len = len(next(iter(pdb_ids)))\n",
    "    for record in SeqIO.parse(fasta_filename, \"fasta\"):\n",
    "        pdb_id = record.id[:id_len] #assuming header begins with lowercase PDB id\n",
    "        if pdb_id in pdb_ids:\n",
    "            SeqIO.write(record, out_fasta_filehandle, \"fasta\")\n",
    "with open(QUERY_SET_PATH, 'w'): pass # clear file content\n",
    "with open(QUERY_SET_PATH, 'a') as out_fasta_file:\n",
    "    extract_sequences(WORK_DIR / 'pdb_seqres.txt', {query_id[:4].lower() + query_id[4:] for query_id in query_ids_set}, out_fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49008abc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb /local/sieg/mmseqs_workdir//pdb_seqres_query.txt /local/sieg/mmseqs_workdir//queryDB \n",
      "\n",
      "MMseqs Version:       \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to queryDB_h: 0h 0m 0s 1ms\n",
      "Time for merging to queryDB: 0h 0m 0s 1ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "createdb /local/sieg/mmseqs_workdir//pdb_seqres.txt /local/sieg/mmseqs_workdir//targetDB \n",
      "\n",
      "MMseqs Version:       \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[==================================================================================\n",
      "Time for merging to targetDB_h: 0h 0m 0s 68ms\n",
      "Time for merging to targetDB: 0h 0m 0s 135ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 1s 255ms\n",
      "Create directory /local/sieg/mmseqs_workdir//tmp\n",
      "search /local/sieg/mmseqs_workdir//queryDB /local/sieg/mmseqs_workdir//targetDB /local/sieg/mmseqs_workdir//results/resultDB /local/sieg/mmseqs_workdir//tmp -a --max-seqs 1000000 \n",
      "\n",
      "MMseqs Version:                        \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Substitution matrix                    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace                          \ttrue\n",
      "Alignment mode                         \t2\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                         \tsubstitution:4.100,context:5.800\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Correlation score weight               \t0\n",
      "Gap open cost                          \taa:11,nucl:5\n",
      "Gap extension cost                     \taa:1,nucl:2\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t6\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                          \taa:21,nucl:5\n",
      "Max results per query                  \t1000000\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask residues probability              \t0.9\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.1\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Use filter only at N seqs              \t0\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0.0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Pseudo count mode                      \t0\n",
      "Gap pseudo count                       \t10\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \tfalse\n",
      "\n",
      "prefilter /local/sieg/mmseqs_workdir//queryDB /local/sieg/mmseqs_workdir//targetDB /local/sieg/mmseqs_workdir//tmp/1500498903493057471/pref_0 --sub-mat aa:blosum62.out,nucl:nucleotide.out --seed-sub-mat aa:VTML80.out,nucl:nucleotide.out -k 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 1000000 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 6 --compressed 0 -v 3 -s 5.7 \n",
      "\n",
      "Query database size: 123 type: Aminoacid\n",
      "Estimated memory consumption: 2G\n",
      "Target database size: 821414 type: Aminoacid\n",
      "Index table k-mer threshold: 112 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 821.41K 5s 638ms\n",
      "Index table: Masked residues: 9342816\n",
      "Index table: fill\n",
      "[=================================================================] 821.41K 6s 634ms\n",
      "Index statistics\n",
      "Entries:          204024933\n",
      "DB size:          1655 MB\n",
      "Avg k-mer size:   3.187890\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t15586\n",
      "    SGVRHM\t14873\n",
      "    VKYPVT\t9169\n",
      "    WNGLGV\t8964\n",
      "    LYLSTV\t8860\n",
      "    GQTVSS\t8408\n",
      "    KDTSST\t7222\n",
      "    SVCLFY\t7202\n",
      "    TVWSLT\t6658\n",
      "    SSLSSV\t6651\n",
      "Time for index table init: 0h 0m 12s 957ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 112\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 123\n",
      "Target db start 1 to 821414\n",
      "[=================================================================] 123 0s 280ms\n",
      "\n",
      "379.346403 k-mers per position\n",
      "331608 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "3472 sequences passed prefiltering per query sequence\n",
      "2581 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 13s 665ms\n",
      "align /local/sieg/mmseqs_workdir//queryDB /local/sieg/mmseqs_workdir//targetDB /local/sieg/mmseqs_workdir//tmp/1500498903493057471/pref_0 /local/sieg/mmseqs_workdir//results/resultDB --sub-mat aa:blosum62.out,nucl:nucleotide.out -a 1 --alignment-mode 2 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 6 --compressed 0 -v 3 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 123 type: Aminoacid\n",
      "Target database size: 821414 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 123 2s 863ms\n",
      "Time for merging to resultDB: 0h 0m 0s 0ms\n",
      "427118 alignments calculated\n",
      "146068 sequence pairs passed the thresholds (0.341985 of overall calculated)\n",
      "1187.544678 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 352ms\n",
      "convertalis --format-mode 4 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov,cigar /local/sieg/mmseqs_workdir//queryDB /local/sieg/mmseqs_workdir//targetDB /local/sieg/mmseqs_workdir//results/resultDB /local/sieg/mmseqs_workdir//results/resultDB.tab \n",
      "\n",
      "MMseqs Version:        \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Substitution matrix    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Alignment format       \t4\n",
      "Format alignment output\tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov,cigar\n",
      "Translation table      \t1\n",
      "Gap open cost          \taa:11,nucl:5\n",
      "Gap extension cost     \taa:1,nucl:2\n",
      "Database output        \tfalse\n",
      "Preload mode           \t0\n",
      "Search type            \t0\n",
      "Threads                \t6\n",
      "Compressed             \t0\n",
      "Verbosity              \t3\n",
      "\n",
      "[=================================================================] 123 0s 62ms\n",
      "Time for merging to resultDB.tab: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 133ms\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "\n",
    "WORK_DIR=\"/local/sieg/mmseqs_workdir/\"\n",
    "\n",
    "mmseqs createdb \"${WORK_DIR}/pdb_seqres_query.txt\" \"${WORK_DIR}/queryDB\"\n",
    "mmseqs createdb \"${WORK_DIR}/pdb_seqres.txt\" \"${WORK_DIR}/targetDB\"\n",
    "\n",
    "mkdir -p \"${WORK_DIR}/results\"\n",
    "\n",
    "mmseqs search \"${WORK_DIR}/queryDB\" \"${WORK_DIR}/targetDB\" \"${WORK_DIR}/results/resultDB\" \"${WORK_DIR}/tmp\" -a --max-seqs 1000000 #--cov-mode 0 --min-seq-id 0 -s 5.7\n",
    " \n",
    "mmseqs convertalis --format-mode 4 --format-output \"query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov,cigar\"\\\n",
    "                   \"${WORK_DIR}/queryDB\" \"${WORK_DIR}/targetDB\" \"${WORK_DIR}/results/resultDB\" \"${WORK_DIR}/results/resultDB.tab\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91aac872",
   "metadata": {},
   "source": [
    "Create a subset of PDBs relevant for us from the all-vs-all experiment on the PDB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "faa8dd4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "MMSEQS_RELEVANT_HITS = WORK_DIR / 'results' / 'resultDB.tab'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8e6407d4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1amk_A</td>\n",
       "      <td>1amk_A</td>\n",
       "      <td>1.000</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1.309000e-162</td>\n",
       "      <td>507</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>251M</td>\n",
       "      <td>1amk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1amk_A</td>\n",
       "      <td>1if2_A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>3.371000e-162</td>\n",
       "      <td>506</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>251M</td>\n",
       "      <td>1if2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1amk_A</td>\n",
       "      <td>2vxn_A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>3.371000e-162</td>\n",
       "      <td>506</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>251M</td>\n",
       "      <td>2vxn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1amk_A</td>\n",
       "      <td>7abx_A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>3.371000e-162</td>\n",
       "      <td>506</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>251M</td>\n",
       "      <td>7abx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1amk_A</td>\n",
       "      <td>7az3_A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>3.371000e-162</td>\n",
       "      <td>506</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>251M</td>\n",
       "      <td>7az3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146063</th>\n",
       "      <td>1nje_A</td>\n",
       "      <td>1b02_A</td>\n",
       "      <td>0.321</td>\n",
       "      <td>314</td>\n",
       "      <td>150</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>304</td>\n",
       "      <td>5</td>\n",
       "      <td>267</td>\n",
       "      <td>2.012000e-43</td>\n",
       "      <td>166</td>\n",
       "      <td>0.956</td>\n",
       "      <td>0.943</td>\n",
       "      <td>15M10D25M1I27M1D19M50I17M1D148M</td>\n",
       "      <td>1b02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146064</th>\n",
       "      <td>1nje_A</td>\n",
       "      <td>3v8h_A</td>\n",
       "      <td>0.322</td>\n",
       "      <td>360</td>\n",
       "      <td>155</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>316</td>\n",
       "      <td>8</td>\n",
       "      <td>327</td>\n",
       "      <td>1.168000e-39</td>\n",
       "      <td>154</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.979</td>\n",
       "      <td>83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M</td>\n",
       "      <td>3v8h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146065</th>\n",
       "      <td>1nje_A</td>\n",
       "      <td>3v8h_B</td>\n",
       "      <td>0.322</td>\n",
       "      <td>360</td>\n",
       "      <td>155</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>316</td>\n",
       "      <td>8</td>\n",
       "      <td>327</td>\n",
       "      <td>1.168000e-39</td>\n",
       "      <td>154</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.979</td>\n",
       "      <td>83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M</td>\n",
       "      <td>3v8h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146066</th>\n",
       "      <td>1nje_A</td>\n",
       "      <td>3v8h_C</td>\n",
       "      <td>0.322</td>\n",
       "      <td>360</td>\n",
       "      <td>155</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>316</td>\n",
       "      <td>8</td>\n",
       "      <td>327</td>\n",
       "      <td>1.168000e-39</td>\n",
       "      <td>154</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.979</td>\n",
       "      <td>83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M</td>\n",
       "      <td>3v8h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146067</th>\n",
       "      <td>1nje_A</td>\n",
       "      <td>3v8h_D</td>\n",
       "      <td>0.322</td>\n",
       "      <td>360</td>\n",
       "      <td>155</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>316</td>\n",
       "      <td>8</td>\n",
       "      <td>327</td>\n",
       "      <td>1.168000e-39</td>\n",
       "      <td>154</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.979</td>\n",
       "      <td>83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M</td>\n",
       "      <td>3v8h</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>146068 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         query  target  fident  alnlen  mismatch  gapopen  qstart  qend   \n",
       "0       1amk_A  1amk_A   1.000     251         0        0       1   251  \\\n",
       "1       1amk_A  1if2_A   0.996     251         1        0       1   251   \n",
       "2       1amk_A  2vxn_A   0.996     251         1        0       1   251   \n",
       "3       1amk_A  7abx_A   0.996     251         1        0       1   251   \n",
       "4       1amk_A  7az3_A   0.996     251         1        0       1   251   \n",
       "...        ...     ...     ...     ...       ...      ...     ...   ...   \n",
       "146063  1nje_A  1b02_A   0.321     314       150        5       3   304   \n",
       "146064  1nje_A  3v8h_A   0.322     360       155        8       6   316   \n",
       "146065  1nje_A  3v8h_B   0.322     360       155        8       6   316   \n",
       "146066  1nje_A  3v8h_C   0.322     360       155        8       6   316   \n",
       "146067  1nje_A  3v8h_D   0.322     360       155        8       6   316   \n",
       "\n",
       "        tstart  tend         evalue  bits   qcov   tcov   \n",
       "0            1   251  1.309000e-162   507  1.000  1.000  \\\n",
       "1            1   251  3.371000e-162   506  1.000  1.000   \n",
       "2            1   251  3.371000e-162   506  1.000  1.000   \n",
       "3            1   251  3.371000e-162   506  1.000  1.000   \n",
       "4            1   251  3.371000e-162   506  1.000  1.000   \n",
       "...        ...   ...            ...   ...    ...    ...   \n",
       "146063       5   267   2.012000e-43   166  0.956  0.943   \n",
       "146064       8   327   1.168000e-39   154  0.984  0.979   \n",
       "146065       8   327   1.168000e-39   154  0.984  0.979   \n",
       "146066       8   327   1.168000e-39   154  0.984  0.979   \n",
       "146067       8   327   1.168000e-39   154  0.984  0.979   \n",
       "\n",
       "                                              cigar target_pdbid  \n",
       "0                                              251M         1amk  \n",
       "1                                              251M         1if2  \n",
       "2                                              251M         2vxn  \n",
       "3                                              251M         7abx  \n",
       "4                                              251M         7az3  \n",
       "...                                             ...          ...  \n",
       "146063              15M10D25M1I27M1D19M50I17M1D148M         1b02  \n",
       "146064  83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M         3v8h  \n",
       "146065  83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M         3v8h  \n",
       "146066  83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M         3v8h  \n",
       "146067  83M2D12M40I13M16D5M18D47M2D80M3D2M5D3M3D26M         3v8h  \n",
       "\n",
       "[146068 rows x 16 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mmseqs = pd.read_csv(MMSEQS_RELEVANT_HITS, sep='\\t')\n",
    "df_mmseqs['target_pdbid'] = df_mmseqs['target'].str[:4]\n",
    "df_mmseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c222d6be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating known mutations of protherm\n",
      "Evaluating known mutations of platinum\n",
      "Evaluating known mutations of thermomutdb\n",
      "Evaluating known mutations of shanthirabalan\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wild_pdb</th>\n",
       "      <th>mut_pdb</th>\n",
       "      <th>wild_aa</th>\n",
       "      <th>mut_aa</th>\n",
       "      <th>wild_seq_num</th>\n",
       "      <th>wild_chain</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBT</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBV</td>\n",
       "      <td>K</td>\n",
       "      <td>A</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBU</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1IOB</td>\n",
       "      <td>1HIB</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>9</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1BNI</td>\n",
       "      <td>1BRH</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>3TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>Y</td>\n",
       "      <td>413</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENP</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENS</td>\n",
       "      <td>E</td>\n",
       "      <td>Q</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENV</td>\n",
       "      <td>S</td>\n",
       "      <td>I</td>\n",
       "      <td>234</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENW</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>234</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1433 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     wild_pdb mut_pdb wild_aa mut_aa wild_seq_num wild_chain         dataset\n",
       "0        2RN2    1RBT       K      G           95          A        protherm\n",
       "1        2RN2    1RBV       K      A           95          A        protherm\n",
       "2        2RN2    1RBU       K      N           95          A        protherm\n",
       "3        1IOB    1HIB       T      G            9          A        protherm\n",
       "4        1BNI    1BRH       L      A           14          A        protherm\n",
       "...       ...     ...     ...    ...          ...        ...             ...\n",
       "1428     4BFL    3TTT       F      Y          413          A  shanthirabalan\n",
       "1429     4BFL    4ENP       E      A          530          A  shanthirabalan\n",
       "1430     4BFL    4ENS       E      Q          530          A  shanthirabalan\n",
       "1431     4BFL    4ENV       S      I          234          A  shanthirabalan\n",
       "1432     4BFL    4ENW       S      N          234          A  shanthirabalan\n",
       "\n",
       "[1433 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# stack all individual benchmark data sets to a single dataframe\n",
    "\n",
    "df_ref_list = []\n",
    "for dataset_name in supported_datasets:\n",
    "    dataset = dataset_collection.get_dataset(dataset_name)\n",
    "            \n",
    "    print(f\"Evaluating known mutations of {dataset.name}\")\n",
    "    df_ref = dataset.read_single_mutations(pdb_mutant_only=True)\n",
    "    \n",
    "    df_ref = df_ref[~df_ref['wild_pdb'].isin(BAD_PDBIDS)]\n",
    "    df_ref = df_ref[~df_ref['mut_pdb'].isin(BAD_PDBIDS)]\n",
    "    \n",
    "    df_ref['dataset'] = dataset_name\n",
    "    \n",
    "    if dataset_name == \"protherm\":\n",
    "        # protherm has no chain identifiers. However, mmseqs needs them. Also its 'A' in all but one case\n",
    "        df_ref[helper.WILD_CHAIN] = 'A'\n",
    "        df_ref.loc[df_ref[helper.WILD_COL] == \"2CI2\", helper.WILD_CHAIN] = 'I'\n",
    "    \n",
    "    df_ref = df_ref[[helper.WILD_COL, helper.MUTANT_COL, helper.WILD_AA, helper.MUT_AA, helper.WILD_SEQ_NUM, helper.WILD_CHAIN, 'dataset']]\n",
    "    df_ref = df_ref.drop_duplicates()\n",
    "    df_ref_list.append(df_ref)\n",
    "df_ref = pd.concat(df_ref_list).reset_index(drop=True)\n",
    "df_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ef90be2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wild_pdb</th>\n",
       "      <th>mut_pdb</th>\n",
       "      <th>wild_aa</th>\n",
       "      <th>mut_aa</th>\n",
       "      <th>wild_seq_num</th>\n",
       "      <th>wild_chain</th>\n",
       "      <th>dataset</th>\n",
       "      <th>mmseqs_wild_ids</th>\n",
       "      <th>mmseqs_mut_id_prefixes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBT</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBV</td>\n",
       "      <td>K</td>\n",
       "      <td>A</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBU</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1IOB</td>\n",
       "      <td>1HIB</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>9</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1iob_A</td>\n",
       "      <td>1hib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1BNI</td>\n",
       "      <td>1BRH</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1bni_A</td>\n",
       "      <td>1brh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>3TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>Y</td>\n",
       "      <td>413</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>3ttt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENP</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4enp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENS</td>\n",
       "      <td>E</td>\n",
       "      <td>Q</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4ens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENV</td>\n",
       "      <td>S</td>\n",
       "      <td>I</td>\n",
       "      <td>234</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4env</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENW</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>234</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4enw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1433 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     wild_pdb mut_pdb wild_aa mut_aa wild_seq_num wild_chain         dataset   \n",
       "0        2RN2    1RBT       K      G           95          A        protherm  \\\n",
       "1        2RN2    1RBV       K      A           95          A        protherm   \n",
       "2        2RN2    1RBU       K      N           95          A        protherm   \n",
       "3        1IOB    1HIB       T      G            9          A        protherm   \n",
       "4        1BNI    1BRH       L      A           14          A        protherm   \n",
       "...       ...     ...     ...    ...          ...        ...             ...   \n",
       "1428     4BFL    3TTT       F      Y          413          A  shanthirabalan   \n",
       "1429     4BFL    4ENP       E      A          530          A  shanthirabalan   \n",
       "1430     4BFL    4ENS       E      Q          530          A  shanthirabalan   \n",
       "1431     4BFL    4ENV       S      I          234          A  shanthirabalan   \n",
       "1432     4BFL    4ENW       S      N          234          A  shanthirabalan   \n",
       "\n",
       "     mmseqs_wild_ids mmseqs_mut_id_prefixes  \n",
       "0             2rn2_A                   1rbt  \n",
       "1             2rn2_A                   1rbv  \n",
       "2             2rn2_A                   1rbu  \n",
       "3             1iob_A                   1hib  \n",
       "4             1bni_A                   1brh  \n",
       "...              ...                    ...  \n",
       "1428          4bfl_A                   3ttt  \n",
       "1429          4bfl_A                   4enp  \n",
       "1430          4bfl_A                   4ens  \n",
       "1431          4bfl_A                   4env  \n",
       "1432          4bfl_A                   4enw  \n",
       "\n",
       "[1433 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add identifiers that can be matched against mmseqs identifiers We use <PDB-id>_<chain-id> for the query and only the pdbid as prefix for the target.\n",
    "df_ref['mmseqs_wild_ids'] = df_ref['wild_pdb'].str.lower() + '_' + df_ref['wild_chain']\n",
    "df_ref['mmseqs_mut_id_prefixes'] = df_ref['mut_pdb'].str.lower() \n",
    "df_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5157185f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The union of all benchmark data sets contains 1041 unique query-chain-target-structure pairs\n",
      "The union of all benchmark data sets contains 1136 unique mutations (w/o chain)\n",
      "The union of all benchmark data sets contains 1137 unique mutations (w chain)\n",
      "The union of all benchmark data sets contains 590 unique mutations (w/o chain; w/o shanthirabalan)\n",
      "The union of all benchmark data sets contains 591 unique mutations (w chain; w/o shanthirabalan)\n"
     ]
    }
   ],
   "source": [
    "df_ref_no_shanthi = df_ref.query('dataset != \"shanthirabalan\"')\n",
    "mut_key_wochain = [helper.WILD_COL, helper.MUTANT_COL, helper.WILD_AA, helper.MUT_AA, helper.WILD_SEQ_NUM]\n",
    "mut_key_wchain = mut_key_wochain + [helper.WILD_CHAIN]\n",
    "\n",
    "nof_unique_query_chain_target_structure_pairs = df_ref.drop_duplicates(['mmseqs_wild_ids', 'mmseqs_mut_id_prefixes']).shape[0]\n",
    "nof_unique_mutations_wo_chain = df_ref.drop_duplicates(mut_key_wochain).shape[0]\n",
    "nof_unique_mutations_w_chain = df_ref.drop_duplicates(mut_key_wchain).shape[0]\n",
    "print(f\"The union of all benchmark data sets contains {nof_unique_query_chain_target_structure_pairs} unique query-chain-target-structure pairs\")\n",
    "print(f\"The union of all benchmark data sets contains {nof_unique_mutations_wo_chain} unique mutations (w/o chain)\")\n",
    "print(f\"The union of all benchmark data sets contains {nof_unique_mutations_w_chain} unique mutations (w chain)\")\n",
    "print(f\"The union of all benchmark data sets contains {df_ref_no_shanthi.drop_duplicates(mut_key_wochain).shape[0]} unique mutations (w/o chain; w/o shanthirabalan)\")\n",
    "print(f\"The union of all benchmark data sets contains {df_ref_no_shanthi.drop_duplicates(mut_key_wchain).shape[0]} unique mutations (w chain; w/o shanthirabalan)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "901122bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91931</th>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>1iph_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1557</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>753M</td>\n",
       "      <td>1iph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91933</th>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>6by0_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1557</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>753M</td>\n",
       "      <td>6by0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91939</th>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>1gge_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1557</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>753M</td>\n",
       "      <td>1gge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91943</th>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>3vu3_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1557</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>753M</td>\n",
       "      <td>3vu3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91944</th>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1557</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>753M</td>\n",
       "      <td>4bfl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124113</th>\n",
       "      <td>2jbz_A</td>\n",
       "      <td>5umw_E</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000968</td>\n",
       "      <td>42</td>\n",
       "      <td>0.126</td>\n",
       "      <td>0.119</td>\n",
       "      <td>18M</td>\n",
       "      <td>5umw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124235</th>\n",
       "      <td>2jbz_A</td>\n",
       "      <td>1t3d_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000968</td>\n",
       "      <td>42</td>\n",
       "      <td>0.126</td>\n",
       "      <td>0.062</td>\n",
       "      <td>18M</td>\n",
       "      <td>1t3d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124251</th>\n",
       "      <td>2jbz_A</td>\n",
       "      <td>4txd_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000968</td>\n",
       "      <td>42</td>\n",
       "      <td>0.126</td>\n",
       "      <td>0.046</td>\n",
       "      <td>18M</td>\n",
       "      <td>4txd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124272</th>\n",
       "      <td>2jbz_A</td>\n",
       "      <td>7blf_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000968</td>\n",
       "      <td>42</td>\n",
       "      <td>0.126</td>\n",
       "      <td>0.040</td>\n",
       "      <td>18M</td>\n",
       "      <td>7blf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124276</th>\n",
       "      <td>2jbz_A</td>\n",
       "      <td>6m9s_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000968</td>\n",
       "      <td>42</td>\n",
       "      <td>0.126</td>\n",
       "      <td>0.037</td>\n",
       "      <td>18M</td>\n",
       "      <td>6m9s</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74380 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         query  target  fident  alnlen  mismatch  gapopen  qstart  qend   \n",
       "91931   4bfl_A  1iph_A     1.0     753         0        0       1   753  \\\n",
       "91933   4bfl_A  6by0_A     1.0     753         0        0       1   753   \n",
       "91939   4bfl_A  1gge_A     1.0     753         0        0       1   753   \n",
       "91943   4bfl_A  3vu3_A     1.0     753         0        0       1   753   \n",
       "91944   4bfl_A  4bfl_A     1.0     753         0        0       1   753   \n",
       "...        ...     ...     ...     ...       ...      ...     ...   ...   \n",
       "124113  2jbz_A  5umw_E     1.0      18         0        0       5    22   \n",
       "124235  2jbz_A  1t3d_A     1.0      18         0        0       5    22   \n",
       "124251  2jbz_A  4txd_A     1.0      18         0        0       5    22   \n",
       "124272  2jbz_A  7blf_A     1.0      18         0        0       5    22   \n",
       "124276  2jbz_A  6m9s_A     1.0      18         0        0       5    22   \n",
       "\n",
       "        tstart  tend    evalue  bits   qcov   tcov cigar target_pdbid  \n",
       "91931        1   753  0.000000  1557  1.000  1.000  753M         1iph  \n",
       "91933        1   753  0.000000  1557  1.000  1.000  753M         6by0  \n",
       "91939        1   753  0.000000  1557  1.000  1.000  753M         1gge  \n",
       "91943        1   753  0.000000  1557  1.000  1.000  753M         3vu3  \n",
       "91944        1   753  0.000000  1557  1.000  1.000  753M         4bfl  \n",
       "...        ...   ...       ...   ...    ...    ...   ...          ...  \n",
       "124113       1    18  0.000968    42  0.126  0.119   18M         5umw  \n",
       "124235       1    18  0.000968    42  0.126  0.062   18M         1t3d  \n",
       "124251       1    18  0.000968    42  0.126  0.046   18M         4txd  \n",
       "124272       1    18  0.000968    42  0.126  0.040   18M         7blf  \n",
       "124276       2    19  0.000968    42  0.126  0.037   18M         6m9s  \n",
       "\n",
       "[74380 rows x 16 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For mmseqs we can only match <pdb-id>_<chain-id> for the query and pdb-id for the target. Let's remove duplicates to make merging easier.\n",
    "df_mmseqs_slim = df_mmseqs.sort_values(['alnlen', 'evalue', 'fident'], ascending=[False, True, False]).drop_duplicates(['query', 'target_pdbid'])\n",
    "df_mmseqs_slim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ada5f9c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wild_pdb</th>\n",
       "      <th>mut_pdb</th>\n",
       "      <th>wild_aa</th>\n",
       "      <th>mut_aa</th>\n",
       "      <th>wild_seq_num</th>\n",
       "      <th>wild_chain</th>\n",
       "      <th>dataset</th>\n",
       "      <th>mmseqs_wild_ids</th>\n",
       "      <th>mmseqs_mut_id_prefixes</th>\n",
       "      <th>query</th>\n",
       "      <th>...</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBT</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbt</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>1.366000e-104</td>\n",
       "      <td>334.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155M</td>\n",
       "      <td>1rbt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBV</td>\n",
       "      <td>K</td>\n",
       "      <td>A</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbv</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>9.961000e-105</td>\n",
       "      <td>334.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155M</td>\n",
       "      <td>1rbv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBU</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbu</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>7.265000e-105</td>\n",
       "      <td>335.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155M</td>\n",
       "      <td>1rbu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1IOB</td>\n",
       "      <td>1HIB</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>9</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1iob_A</td>\n",
       "      <td>1hib</td>\n",
       "      <td>1iob_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>1.384000e-95</td>\n",
       "      <td>308.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>153M</td>\n",
       "      <td>1hib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1BNI</td>\n",
       "      <td>1BRH</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1bni_A</td>\n",
       "      <td>1brh</td>\n",
       "      <td>1bni_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.979000e-70</td>\n",
       "      <td>233.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>110M</td>\n",
       "      <td>1brh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>3TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>Y</td>\n",
       "      <td>413</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>3ttt</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753M</td>\n",
       "      <td>3ttt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENP</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4enp</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1555.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753M</td>\n",
       "      <td>4enp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENS</td>\n",
       "      <td>E</td>\n",
       "      <td>Q</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4ens</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753M</td>\n",
       "      <td>4ens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENV</td>\n",
       "      <td>S</td>\n",
       "      <td>I</td>\n",
       "      <td>234</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4env</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1555.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753M</td>\n",
       "      <td>4env</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>4BFL</td>\n",
       "      <td>4ENW</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>234</td>\n",
       "      <td>A</td>\n",
       "      <td>shanthirabalan</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>4enw</td>\n",
       "      <td>4bfl_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>753M</td>\n",
       "      <td>4enw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1433 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     wild_pdb mut_pdb wild_aa mut_aa wild_seq_num wild_chain         dataset   \n",
       "0        2RN2    1RBT       K      G           95          A        protherm  \\\n",
       "1        2RN2    1RBV       K      A           95          A        protherm   \n",
       "2        2RN2    1RBU       K      N           95          A        protherm   \n",
       "3        1IOB    1HIB       T      G            9          A        protherm   \n",
       "4        1BNI    1BRH       L      A           14          A        protherm   \n",
       "...       ...     ...     ...    ...          ...        ...             ...   \n",
       "1428     4BFL    3TTT       F      Y          413          A  shanthirabalan   \n",
       "1429     4BFL    4ENP       E      A          530          A  shanthirabalan   \n",
       "1430     4BFL    4ENS       E      Q          530          A  shanthirabalan   \n",
       "1431     4BFL    4ENV       S      I          234          A  shanthirabalan   \n",
       "1432     4BFL    4ENW       S      N          234          A  shanthirabalan   \n",
       "\n",
       "     mmseqs_wild_ids mmseqs_mut_id_prefixes   query  ... qstart   qend   \n",
       "0             2rn2_A                   1rbt  2rn2_A  ...    1.0  155.0  \\\n",
       "1             2rn2_A                   1rbv  2rn2_A  ...    1.0  155.0   \n",
       "2             2rn2_A                   1rbu  2rn2_A  ...    1.0  155.0   \n",
       "3             1iob_A                   1hib  1iob_A  ...    1.0  153.0   \n",
       "4             1bni_A                   1brh  1bni_A  ...    1.0  110.0   \n",
       "...              ...                    ...     ...  ...    ...    ...   \n",
       "1428          4bfl_A                   3ttt  4bfl_A  ...    1.0  753.0   \n",
       "1429          4bfl_A                   4enp  4bfl_A  ...    1.0  753.0   \n",
       "1430          4bfl_A                   4ens  4bfl_A  ...    1.0  753.0   \n",
       "1431          4bfl_A                   4env  4bfl_A  ...    1.0  753.0   \n",
       "1432          4bfl_A                   4enw  4bfl_A  ...    1.0  753.0   \n",
       "\n",
       "      tstart   tend         evalue    bits  qcov  tcov  cigar  target_pdbid  \n",
       "0        1.0  155.0  1.366000e-104   334.0   1.0   1.0   155M          1rbt  \n",
       "1        1.0  155.0  9.961000e-105   334.0   1.0   1.0   155M          1rbv  \n",
       "2        1.0  155.0  7.265000e-105   335.0   1.0   1.0   155M          1rbu  \n",
       "3        1.0  153.0   1.384000e-95   308.0   1.0   1.0   153M          1hib  \n",
       "4        1.0  110.0   1.979000e-70   233.0   1.0   1.0   110M          1brh  \n",
       "...      ...    ...            ...     ...   ...   ...    ...           ...  \n",
       "1428     1.0  753.0   0.000000e+00  1556.0   1.0   1.0   753M          3ttt  \n",
       "1429     1.0  753.0   0.000000e+00  1555.0   1.0   1.0   753M          4enp  \n",
       "1430     1.0  753.0   0.000000e+00  1556.0   1.0   1.0   753M          4ens  \n",
       "1431     1.0  753.0   0.000000e+00  1555.0   1.0   1.0   753M          4env  \n",
       "1432     1.0  753.0   0.000000e+00  1556.0   1.0   1.0   753M          4enw  \n",
       "\n",
       "[1433 rows x 25 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ref_w_mmseqs = df_ref.merge(df_mmseqs_slim, left_on=['mmseqs_wild_ids', 'mmseqs_mut_id_prefixes'], right_on=['query', 'target_pdbid'], how='left')\n",
    "assert df_ref_w_mmseqs.shape[0] == df_ref.shape[0]\n",
    "df_ref_w_mmseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4da78b0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wild_pdb</th>\n",
       "      <th>mut_pdb</th>\n",
       "      <th>wild_aa</th>\n",
       "      <th>mut_aa</th>\n",
       "      <th>wild_seq_num</th>\n",
       "      <th>wild_chain</th>\n",
       "      <th>dataset</th>\n",
       "      <th>mmseqs_wild_ids</th>\n",
       "      <th>mmseqs_mut_id_prefixes</th>\n",
       "      <th>query</th>\n",
       "      <th>...</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1BNI</td>\n",
       "      <td>1BGD</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>96</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1bni_A</td>\n",
       "      <td>1bgd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBN</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1YAG</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>89</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1yag</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA2</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>74</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>110</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>3PCA</td>\n",
       "      <td>1YKL</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>408</td>\n",
       "      <td>I</td>\n",
       "      <td>platinum</td>\n",
       "      <td>3pca_I</td>\n",
       "      <td>1ykl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>3PCA</td>\n",
       "      <td>1YKP</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>408</td>\n",
       "      <td>I</td>\n",
       "      <td>platinum</td>\n",
       "      <td>3pca_I</td>\n",
       "      <td>1ykp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>1R2Y</td>\n",
       "      <td>3GQ9</td>\n",
       "      <td>R</td>\n",
       "      <td>E</td>\n",
       "      <td>244</td>\n",
       "      <td>A</td>\n",
       "      <td>thermomutdb</td>\n",
       "      <td>1r2y_A</td>\n",
       "      <td>3gq9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    wild_pdb mut_pdb wild_aa mut_aa wild_seq_num wild_chain      dataset   \n",
       "14      1BNI    1BGD       I      V           96          A     protherm  \\\n",
       "54      2RN2    1RBN       K      N           95          A     protherm   \n",
       "262     1LZ1    1YAG       I      V           89          A     protherm   \n",
       "278     1LZ1    1GA2       V      I            2          A     protherm   \n",
       "279     1LZ1    1GA0       V      L            2          A     protherm   \n",
       "295     1LZ1    1GA0       V      L           74          A     protherm   \n",
       "300     1LZ1    1GA0       V      L          110          A     protherm   \n",
       "371     3PCA    1YKL       Y      C          408          I     platinum   \n",
       "372     3PCA    1YKP       Y      H          408          I     platinum   \n",
       "816     1R2Y    3GQ9       R      E          244          A  thermomutdb   \n",
       "\n",
       "    mmseqs_wild_ids mmseqs_mut_id_prefixes query  ... qstart  qend  tstart   \n",
       "14           1bni_A                   1bgd   NaN  ...    NaN   NaN     NaN  \\\n",
       "54           2rn2_A                   1rbn   NaN  ...    NaN   NaN     NaN   \n",
       "262          1lz1_A                   1yag   NaN  ...    NaN   NaN     NaN   \n",
       "278          1lz1_A                   1ga2   NaN  ...    NaN   NaN     NaN   \n",
       "279          1lz1_A                   1ga0   NaN  ...    NaN   NaN     NaN   \n",
       "295          1lz1_A                   1ga0   NaN  ...    NaN   NaN     NaN   \n",
       "300          1lz1_A                   1ga0   NaN  ...    NaN   NaN     NaN   \n",
       "371          3pca_I                   1ykl   NaN  ...    NaN   NaN     NaN   \n",
       "372          3pca_I                   1ykp   NaN  ...    NaN   NaN     NaN   \n",
       "816          1r2y_A                   3gq9   NaN  ...    NaN   NaN     NaN   \n",
       "\n",
       "     tend  evalue  bits  qcov  tcov  cigar  target_pdbid  \n",
       "14    NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "54    NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "262   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "278   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "279   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "295   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "300   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "371   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "372   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "816   NaN     NaN   NaN   NaN   NaN    NaN           NaN  \n",
       "\n",
       "[10 rows x 25 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show what mmseqs did not find\n",
    "df_mmseqs_not_found = df_ref_w_mmseqs.query('query != query').drop_duplicates(['wild_pdb', 'mut_pdb', 'wild_aa', 'mut_aa', 'wild_seq_num', 'wild_chain'])\n",
    "df_mmseqs_not_found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3ebed42f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wild_pdb</th>\n",
       "      <th>mut_pdb</th>\n",
       "      <th>wild_aa</th>\n",
       "      <th>mut_aa</th>\n",
       "      <th>wild_seq_num</th>\n",
       "      <th>wild_chain_x</th>\n",
       "      <th>dataset</th>\n",
       "      <th>mmseqs_wild_ids</th>\n",
       "      <th>mmseqs_mut_id_prefixes</th>\n",
       "      <th>query</th>\n",
       "      <th>...</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "      <th>pdb_wild</th>\n",
       "      <th>pdb_mutant</th>\n",
       "      <th>wild_chain_y</th>\n",
       "      <th>seq_num</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>reason</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1BNI</td>\n",
       "      <td>1BGD</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>96</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1bni_A</td>\n",
       "      <td>1bgd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1BNI</td>\n",
       "      <td>1BGD</td>\n",
       "      <td>A</td>\n",
       "      <td>96</td>\n",
       "      <td>V</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBN</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>2rn2_A</td>\n",
       "      <td>1rbn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2RN2</td>\n",
       "      <td>1RBN</td>\n",
       "      <td>A</td>\n",
       "      <td>95</td>\n",
       "      <td>N</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1YAG</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>89</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1yag</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1YAG</td>\n",
       "      <td>A</td>\n",
       "      <td>89</td>\n",
       "      <td>V</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA2</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA2</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>I</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>L</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>74</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>A</td>\n",
       "      <td>74</td>\n",
       "      <td>L</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>110</td>\n",
       "      <td>A</td>\n",
       "      <td>protherm</td>\n",
       "      <td>1lz1_A</td>\n",
       "      <td>1ga0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1LZ1</td>\n",
       "      <td>1GA0</td>\n",
       "      <td>A</td>\n",
       "      <td>110</td>\n",
       "      <td>L</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3PCA</td>\n",
       "      <td>1YKL</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>408</td>\n",
       "      <td>I</td>\n",
       "      <td>platinum</td>\n",
       "      <td>3pca_I</td>\n",
       "      <td>1ykl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3PCA</td>\n",
       "      <td>1YKP</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>408</td>\n",
       "      <td>I</td>\n",
       "      <td>platinum</td>\n",
       "      <td>3pca_I</td>\n",
       "      <td>1ykp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1R2Y</td>\n",
       "      <td>3GQ9</td>\n",
       "      <td>R</td>\n",
       "      <td>E</td>\n",
       "      <td>244</td>\n",
       "      <td>A</td>\n",
       "      <td>thermomutdb</td>\n",
       "      <td>1r2y_A</td>\n",
       "      <td>3gq9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1R2Y</td>\n",
       "      <td>3GQ9</td>\n",
       "      <td>A</td>\n",
       "      <td>244</td>\n",
       "      <td>E</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  wild_pdb mut_pdb wild_aa mut_aa wild_seq_num wild_chain_x      dataset   \n",
       "0     1BNI    1BGD       I      V           96            A     protherm  \\\n",
       "1     2RN2    1RBN       K      N           95            A     protherm   \n",
       "2     1LZ1    1YAG       I      V           89            A     protherm   \n",
       "3     1LZ1    1GA2       V      I            2            A     protherm   \n",
       "4     1LZ1    1GA0       V      L            2            A     protherm   \n",
       "5     1LZ1    1GA0       V      L           74            A     protherm   \n",
       "6     1LZ1    1GA0       V      L          110            A     protherm   \n",
       "7     3PCA    1YKL       Y      C          408            I     platinum   \n",
       "8     3PCA    1YKP       Y      H          408            I     platinum   \n",
       "9     1R2Y    3GQ9       R      E          244            A  thermomutdb   \n",
       "\n",
       "  mmseqs_wild_ids mmseqs_mut_id_prefixes query  ... tcov  cigar  target_pdbid   \n",
       "0          1bni_A                   1bgd   NaN  ...  NaN    NaN           NaN  \\\n",
       "1          2rn2_A                   1rbn   NaN  ...  NaN    NaN           NaN   \n",
       "2          1lz1_A                   1yag   NaN  ...  NaN    NaN           NaN   \n",
       "3          1lz1_A                   1ga2   NaN  ...  NaN    NaN           NaN   \n",
       "4          1lz1_A                   1ga0   NaN  ...  NaN    NaN           NaN   \n",
       "5          1lz1_A                   1ga0   NaN  ...  NaN    NaN           NaN   \n",
       "6          1lz1_A                   1ga0   NaN  ...  NaN    NaN           NaN   \n",
       "7          3pca_I                   1ykl   NaN  ...  NaN    NaN           NaN   \n",
       "8          3pca_I                   1ykp   NaN  ...  NaN    NaN           NaN   \n",
       "9          1r2y_A                   3gq9   NaN  ...  NaN    NaN           NaN   \n",
       "\n",
       "   pdb_wild  pdb_mutant  wild_chain_y  seq_num  mutation_aa   \n",
       "0      1BNI        1BGD             A       96            V  \\\n",
       "1      2RN2        1RBN             A       95            N   \n",
       "2      1LZ1        1YAG             A       89            V   \n",
       "3      1LZ1        1GA2             A        2            I   \n",
       "4      1LZ1        1GA0             A        2            L   \n",
       "5      1LZ1        1GA0             A       74            L   \n",
       "6      1LZ1        1GA0             A      110            L   \n",
       "7       NaN         NaN           NaN      NaN          NaN   \n",
       "8       NaN         NaN           NaN      NaN          NaN   \n",
       "9      1R2Y        3GQ9             A      244            E   \n",
       "\n",
       "               reason     _merge  \n",
       "0  different proteins       both  \n",
       "1  different proteins       both  \n",
       "2  different proteins       both  \n",
       "3  different proteins       both  \n",
       "4  different proteins       both  \n",
       "5  different proteins       both  \n",
       "6  different proteins       both  \n",
       "7                 NaN  left_only  \n",
       "8                 NaN  left_only  \n",
       "9  different proteins       both  \n",
       "\n",
       "[10 rows x 32 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare benchmark pairs mmseqs not find with the set of problematic dataset mutation identified previousely\n",
    "df_mmseqs_not_found.merge(df_problematic, left_on=[helper.WILD_COL, helper.MUTANT_COL, helper.WILD_AA, helper.MUT_AA, helper.WILD_SEQ_NUM],\n",
    "                         right_on= ['pdb_wild', 'pdb_mutant', 'wild_aa', 'mutation_aa', 'seq_num'], how='left', indicator=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78bc7719",
   "metadata": {},
   "source": [
    "As we can see all pairs not found by MMseqs are in the problematic set. Only 3PCA-1YKL and 3PCA-1YKP is not found because of the custom chain identifiers in platinum. (see the _merge column)\n",
    "\n",
    "However, when we search for 3pca_A in the mmseqs results we find the expected hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "11cff5ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54455</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54461</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54467</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_E</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54472</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54476</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_I</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54481</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_K</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54539</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54543</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54549</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_E</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54555</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54560</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_I</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54565</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_K</td>\n",
       "      <td>1.0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>9.231000e-135</td>\n",
       "      <td>424</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>200M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54779</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_B</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>1.614000e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54782</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_D</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>1.614000e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54783</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_F</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>1.614000e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54784</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_H</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>1.614000e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54785</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_J</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>1.614000e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54786</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykp_L</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>1.614000e-15</td>\n",
       "      <td>79</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54790</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_L</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>5.539000e-15</td>\n",
       "      <td>78</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54809</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_B</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>5.539000e-15</td>\n",
       "      <td>78</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54810</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_D</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>5.539000e-15</td>\n",
       "      <td>78</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54811</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_F</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>5.539000e-15</td>\n",
       "      <td>78</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54812</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_H</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>5.539000e-15</td>\n",
       "      <td>78</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54813</th>\n",
       "      <td>3pca_A</td>\n",
       "      <td>1ykl_J</td>\n",
       "      <td>0.3</td>\n",
       "      <td>206</td>\n",
       "      <td>122</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>235</td>\n",
       "      <td>5.539000e-15</td>\n",
       "      <td>78</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.828</td>\n",
       "      <td>11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M</td>\n",
       "      <td>1ykl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        query  target  fident  alnlen  mismatch  gapopen  qstart  qend   \n",
       "54455  3pca_A  1ykp_A     1.0     200         0        0       1   200  \\\n",
       "54461  3pca_A  1ykp_C     1.0     200         0        0       1   200   \n",
       "54467  3pca_A  1ykp_E     1.0     200         0        0       1   200   \n",
       "54472  3pca_A  1ykp_G     1.0     200         0        0       1   200   \n",
       "54476  3pca_A  1ykp_I     1.0     200         0        0       1   200   \n",
       "54481  3pca_A  1ykp_K     1.0     200         0        0       1   200   \n",
       "54539  3pca_A  1ykl_A     1.0     200         0        0       1   200   \n",
       "54543  3pca_A  1ykl_C     1.0     200         0        0       1   200   \n",
       "54549  3pca_A  1ykl_E     1.0     200         0        0       1   200   \n",
       "54555  3pca_A  1ykl_G     1.0     200         0        0       1   200   \n",
       "54560  3pca_A  1ykl_I     1.0     200         0        0       1   200   \n",
       "54565  3pca_A  1ykl_K     1.0     200         0        0       1   200   \n",
       "54779  3pca_A  1ykp_B     0.3     206       122        9       5   197   \n",
       "54782  3pca_A  1ykp_D     0.3     206       122        9       5   197   \n",
       "54783  3pca_A  1ykp_F     0.3     206       122        9       5   197   \n",
       "54784  3pca_A  1ykp_H     0.3     206       122        9       5   197   \n",
       "54785  3pca_A  1ykp_J     0.3     206       122        9       5   197   \n",
       "54786  3pca_A  1ykp_L     0.3     206       122        9       5   197   \n",
       "54790  3pca_A  1ykl_L     0.3     206       122        9       5   197   \n",
       "54809  3pca_A  1ykl_B     0.3     206       122        9       5   197   \n",
       "54810  3pca_A  1ykl_D     0.3     206       122        9       5   197   \n",
       "54811  3pca_A  1ykl_F     0.3     206       122        9       5   197   \n",
       "54812  3pca_A  1ykl_H     0.3     206       122        9       5   197   \n",
       "54813  3pca_A  1ykl_J     0.3     206       122        9       5   197   \n",
       "\n",
       "       tstart  tend         evalue  bits   qcov   tcov   \n",
       "54455       1   200  9.231000e-135   424  1.000  1.000  \\\n",
       "54461       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54467       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54472       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54476       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54481       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54539       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54543       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54549       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54555       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54560       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54565       1   200  9.231000e-135   424  1.000  1.000   \n",
       "54779      39   235   1.614000e-15    79  0.965  0.828   \n",
       "54782      39   235   1.614000e-15    79  0.965  0.828   \n",
       "54783      39   235   1.614000e-15    79  0.965  0.828   \n",
       "54784      39   235   1.614000e-15    79  0.965  0.828   \n",
       "54785      39   235   1.614000e-15    79  0.965  0.828   \n",
       "54786      39   235   1.614000e-15    79  0.965  0.828   \n",
       "54790      39   235   5.539000e-15    78  0.965  0.828   \n",
       "54809      39   235   5.539000e-15    78  0.965  0.828   \n",
       "54810      39   235   5.539000e-15    78  0.965  0.828   \n",
       "54811      39   235   5.539000e-15    78  0.965  0.828   \n",
       "54812      39   235   5.539000e-15    78  0.965  0.828   \n",
       "54813      39   235   5.539000e-15    78  0.965  0.828   \n",
       "\n",
       "                                                cigar target_pdbid  \n",
       "54455                                            200M         1ykp  \n",
       "54461                                            200M         1ykp  \n",
       "54467                                            200M         1ykp  \n",
       "54472                                            200M         1ykp  \n",
       "54476                                            200M         1ykp  \n",
       "54481                                            200M         1ykp  \n",
       "54539                                            200M         1ykl  \n",
       "54543                                            200M         1ykl  \n",
       "54549                                            200M         1ykl  \n",
       "54555                                            200M         1ykl  \n",
       "54560                                            200M         1ykl  \n",
       "54565                                            200M         1ykl  \n",
       "54779  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykp  \n",
       "54782  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykp  \n",
       "54783  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykp  \n",
       "54784  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykp  \n",
       "54785  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykp  \n",
       "54786  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykp  \n",
       "54790  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykl  \n",
       "54809  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykl  \n",
       "54810  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykl  \n",
       "54811  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykl  \n",
       "54812  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykl  \n",
       "54813  11M1D7M7I16M1D35M3D3M2D17M1I13M2D37M1I21M4D24M         1ykl  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mmseqs.query('(query == \"3pca_A\" and target_pdbid == \"1ykl\") or (query == \"3pca_A\" and target_pdbid == \"1ykp\")')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b5d3277d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MMseqs successfully retrieved 99.42% query_chain_target_structures pairs\n",
      "MMseqs missed 6 query_chain_target_structures pairs\n"
     ]
    }
   ],
   "source": [
    "nof_query_chain_target_structures_with_successful_mmseqs_hit = df_ref_w_mmseqs.query('query == query').drop_duplicates(['mmseqs_wild_ids', 'mmseqs_mut_id_prefixes']).shape[0]\n",
    "nof_unique_query_chain_target_structure_pairs_adjusted = nof_unique_query_chain_target_structure_pairs - 2 # adjust for 3pca_I in platinum\n",
    "print(f'MMseqs successfully retrieved {100*(nof_query_chain_target_structures_with_successful_mmseqs_hit / nof_unique_query_chain_target_structure_pairs_adjusted):.2f}% query_chain_target_structures pairs')\n",
    "print(f'MMseqs missed {nof_unique_query_chain_target_structure_pairs_adjusted - nof_query_chain_target_structures_with_successful_mmseqs_hit} query_chain_target_structures pairs')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ef8b079",
   "metadata": {},
   "source": [
    "## Compare MMseqs results to 'problematic_single_mutations' \n",
    "\n",
    "MMseqs hits include hits that we don't want to find because of local differences in the 3D env and generally errors in the annotations of the datasets "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "de6a0210",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1033"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nof_query_chain_target_structures_with_successful_mmseqs_hit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "87f5f3f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MMseqs reported 9.10% 'problematic mutation' query_chain_target_structures pairs\n",
      "nof_problematic_chain_pairs_reported_by_mmseqs= 94\n"
     ]
    }
   ],
   "source": [
    "df_ref_w_mmseqs_prob = df_ref_w_mmseqs.merge(df_problematic, left_on=[helper.WILD_COL, helper.MUTANT_COL, helper.WILD_AA, helper.MUT_AA, helper.WILD_SEQ_NUM, helper.WILD_CHAIN],\n",
    "                         right_on= ['pdb_wild', 'pdb_mutant', 'wild_aa', 'mutation_aa', 'seq_num', 'wild_chain'], how='left', indicator=True).query('_merge == \"both\" and query== query')\n",
    "nof_problematic_chain_pairs_reported_by_mmseqs = df_ref_w_mmseqs_prob.drop_duplicates([\"mmseqs_wild_ids\", \"mmseqs_mut_id_prefixes\"]).shape[0]\n",
    "print(f'MMseqs reported {100*(nof_problematic_chain_pairs_reported_by_mmseqs / nof_query_chain_target_structures_with_successful_mmseqs_hit):.2f}% \\'problematic mutation\\' query_chain_target_structures pairs')\n",
    "print('nof_problematic_chain_pairs_reported_by_mmseqs=', nof_problematic_chain_pairs_reported_by_mmseqs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f6c84e67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nof_problematic_mutations_reported_by_mmseqs= 178\n",
      "these are 15.67% of all mutations\n"
     ]
    }
   ],
   "source": [
    "nof_problematic_mutations_reported_by_mmseqs = df_ref_w_mmseqs_prob.drop_duplicates([helper.WILD_COL, helper.MUTANT_COL, helper.WILD_AA, helper.MUT_AA, helper.WILD_SEQ_NUM, helper.WILD_CHAIN]).shape[0]\n",
    "print('nof_problematic_mutations_reported_by_mmseqs=', nof_problematic_mutations_reported_by_mmseqs)\n",
    "print(f'these are {100*(nof_problematic_mutations_reported_by_mmseqs / nof_unique_mutations_wo_chain):.2f}% of all mutations')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e88b59f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wild_pdb</th>\n",
       "      <th>mut_pdb</th>\n",
       "      <th>wild_aa</th>\n",
       "      <th>mut_aa</th>\n",
       "      <th>wild_seq_num</th>\n",
       "      <th>wild_chain</th>\n",
       "      <th>dataset</th>\n",
       "      <th>mmseqs_wild_ids</th>\n",
       "      <th>mmseqs_mut_id_prefixes</th>\n",
       "      <th>query</th>\n",
       "      <th>...</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>cigar</th>\n",
       "      <th>target_pdbid</th>\n",
       "      <th>pdb_wild</th>\n",
       "      <th>pdb_mutant</th>\n",
       "      <th>seq_num</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>reason</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>1PGA</td>\n",
       "      <td>6HKA</td>\n",
       "      <td>A</td>\n",
       "      <td>L</td>\n",
       "      <td>34</td>\n",
       "      <td>A</td>\n",
       "      <td>thermomutdb</td>\n",
       "      <td>1pga_A</td>\n",
       "      <td>6hka</td>\n",
       "      <td>1pga_A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.56</td>\n",
       "      <td>56M</td>\n",
       "      <td>6hka</td>\n",
       "      <td>1PGA</td>\n",
       "      <td>6HKA</td>\n",
       "      <td>34</td>\n",
       "      <td>L</td>\n",
       "      <td>different proteins</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    wild_pdb mut_pdb wild_aa mut_aa wild_seq_num wild_chain      dataset   \n",
       "833     1PGA    6HKA       A      L           34          A  thermomutdb  \\\n",
       "\n",
       "    mmseqs_wild_ids mmseqs_mut_id_prefixes   query  ... qcov  tcov  cigar   \n",
       "833          1pga_A                   6hka  1pga_A  ...  1.0  0.56    56M  \\\n",
       "\n",
       "     target_pdbid  pdb_wild  pdb_mutant  seq_num  mutation_aa   \n",
       "833          6hka      1PGA        6HKA       34            L  \\\n",
       "\n",
       "                 reason  _merge  \n",
       "833  different proteins    both  \n",
       "\n",
       "[1 rows x 31 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ref_w_mmseqs_prob.query('reason == \"different proteins\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d61bdff4",
   "metadata": {},
   "source": [
    "## MMseqs reports \"different\" protein in one case\n",
    "\n",
    "TMAlign and MicroMiner don't report the mutation \"1PGA\t6HKA\tA\tA\t34\tL\" from thermomutdb. But MMseqs reports a hit for 1pga_A with 6hka_A.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b56ca986",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">1pga_A mol:protein length:56  PROTEIN G\n",
      "MTYKLILNGKTLKGETTTEAVDAATAEKVFKQYANDNGVDGEWTYDDATKTFTVTE\n",
      ">6hka_A mol:protein length:100  Immunoglobulin G-binding protein G,Serine-protein kinase ATM\n",
      "MQYKLALNGKTLKGETTTEAVDAATAEKVFKQYANDNGVDGEWTYDDATKTFTVTELVPRGSDDDDKTVLSVGGQVNLLIQQAIDPKNLSRLFPGWKAWV\n"
     ]
    }
   ],
   "source": [
    "# look at the seqres fasta sequences of both files\n",
    "!grep -A1 1pga /local/sieg/mmseqs_workdir/pdb_seqres.txt\n",
    "!grep -A1 6hka /local/sieg/mmseqs_workdir/pdb_seqres.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "39c4369a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '1pga'...\n",
      "Downloading PDB structure '6hka'...\n",
      "1pga MTYKLILNGKTLKGETTTEAVDAATAEKVFKQYANDNGVDGEWTYDDATKTFTVTE\n",
      "6hka TVLSVGGQVNLLIQQAIDPKNLSRLFPGWKAWV\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import PDBList, PDBParser, Polypeptide\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "def download_and_extract(pdb_id, chain_id):\n",
    "    pdbl = PDBList()\n",
    "    pdb_file = pdbl.retrieve_pdb_file(pdb_id, pdir=str(WORK_DIR.resolve()), file_format=\"pdb\")\n",
    "\n",
    "    parser = PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure(pdb_id, pdb_file)\n",
    "\n",
    "    atom_sequence = \"\"\n",
    "    models = list(structure.get_models())\n",
    "    model = models[0]\n",
    "    for chain in model:\n",
    "        if chain.id != chain_id:\n",
    "            continue\n",
    "        for residue in chain:\n",
    "            for atom in residue:\n",
    "                if atom.get_id() == 'CA':\n",
    "                    atom_sequence += Polypeptide.three_to_one(residue.get_resname())\n",
    "\n",
    "    return atom_sequence\n",
    "\n",
    "def write_to_fasta(sequence, entry_id, out_dir):\n",
    "    fasta_record = SeqRecord(Seq(sequence), id=entry_id, description=f\"{entry_id} atom sequence\")\n",
    "    fasta_file = f\"{entry_id}.fasta\"\n",
    "    with open(out_dir / fasta_file, \"w\") as output_handle:\n",
    "        output_handle.write(fasta_record.format(\"fasta\"))  \n",
    "\n",
    "atomseq_1pga = download_and_extract('1pga', 'A')\n",
    "atomseq_6hka = download_and_extract('6hka', 'A')\n",
    "\n",
    "write_to_fasta(atomseq_1pga, '1pga_A', WORK_DIR)\n",
    "write_to_fasta(atomseq_6hka, '6hka_A', WORK_DIR)\n",
    " \n",
    "print('1pga', atomseq_1pga)\n",
    "print('6hka', atomseq_6hka)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9472f938",
   "metadata": {},
   "source": [
    "As we can see above while the atom sequence of 1pga is identical to the seqres sequence, the atom sequence extracted from 6hka is very different and contains only the second half of the sequence  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6ff3427f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb /local/sieg/mmseqs_workdir//1pga_A.fasta /local/sieg/mmseqs_workdir//1pga_A_DB \n",
      "\n",
      "MMseqs Version:       \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to 1pga_A_DB_h: 0h 0m 0s 1ms\n",
      "Time for merging to 1pga_A_DB: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "createdb /local/sieg/mmseqs_workdir//6hka_A.fasta /local/sieg/mmseqs_workdir//6hka_A_DB \n",
      "\n",
      "MMseqs Version:       \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to 6hka_A_DB_h: 0h 0m 0s 0ms\n",
      "Time for merging to 6hka_A_DB: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "search /local/sieg/mmseqs_workdir//1pga_A_DB /local/sieg/mmseqs_workdir//6hka_A_DB /local/sieg/mmseqs_workdir//results/1pga_A_6hka_A_resultDB /local/sieg/mmseqs_workdir//tmp -a --max-seqs 1000000 \n",
      "\n",
      "MMseqs Version:                        \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Substitution matrix                    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace                          \ttrue\n",
      "Alignment mode                         \t2\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                         \tsubstitution:4.100,context:5.800\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Correlation score weight               \t0\n",
      "Gap open cost                          \taa:11,nucl:5\n",
      "Gap extension cost                     \taa:1,nucl:2\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t6\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                          \taa:21,nucl:5\n",
      "Max results per query                  \t1000000\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask residues probability              \t0.9\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.1\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Use filter only at N seqs              \t0\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0.0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Pseudo count mode                      \t0\n",
      "Gap pseudo count                       \t10\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \tfalse\n",
      "\n",
      "prefilter /local/sieg/mmseqs_workdir//1pga_A_DB /local/sieg/mmseqs_workdir//6hka_A_DB /local/sieg/mmseqs_workdir//tmp/8916268785369259718/pref_0 --sub-mat aa:blosum62.out,nucl:nucleotide.out --seed-sub-mat aa:VTML80.out,nucl:nucleotide.out -k 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 1000000 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 6 --compressed 0 -v 3 -s 5.7 \n",
      "\n",
      "Query database size: 1 type: Aminoacid\n",
      "Estimated memory consumption: 977M\n",
      "Target database size: 1 type: Aminoacid\n",
      "Index table k-mer threshold: 112 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1 0s 4ms\n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 1 0s 0ms\n",
      "Index statistics\n",
      "Entries:          24\n",
      "DB size:          488 MB\n",
      "Avg k-mer size:   0.000000\n",
      "Top 10 k-mers\n",
      "    LSLPKA\t1\n",
      "    GQNLQA\t1\n",
      "    VNLQID\t1\n",
      "    IDKLLF\t1\n",
      "    PKLRPG\t1\n",
      "    QVLIAI\t1\n",
      "    SVGVLI\t1\n",
      "    LLQAPK\t1\n",
      "    NLRFWK\t1\n",
      "    LSGQLL\t1\n",
      "Time for index table init: 0h 0m 0s 405ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 112\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1\n",
      "Target db start 1 to 1\n",
      "[=================================================================] 1 0s 0ms\n",
      "\n",
      "120.267857 k-mers per position\n",
      "0 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "0 sequences passed prefiltering per query sequence\n",
      "0 median result list length\n",
      "1 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 786ms\n",
      "align /local/sieg/mmseqs_workdir//1pga_A_DB /local/sieg/mmseqs_workdir//6hka_A_DB /local/sieg/mmseqs_workdir//tmp/8916268785369259718/pref_0 /local/sieg/mmseqs_workdir//results/1pga_A_6hka_A_resultDB --sub-mat aa:blosum62.out,nucl:nucleotide.out -a 1 --alignment-mode 2 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 6 --compressed 0 -v 3 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1 type: Aminoacid\n",
      "Target database size: 1 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1 0s 0ms\n",
      "Time for merging to 1pga_A_6hka_A_resultDB: 0h 0m 0s 0ms\n",
      "0 alignments calculated\n",
      "0 sequence pairs passed the thresholds\n",
      "0.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "convertalis --format-mode 4 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov,cigar /local/sieg/mmseqs_workdir//1pga_A_DB /local/sieg/mmseqs_workdir//6hka_A_DB /local/sieg/mmseqs_workdir//results/1pga_A_6hka_A_resultDB /local/sieg/mmseqs_workdir//results/1pga_A_6hka_A_resultDB.tab \n",
      "\n",
      "MMseqs Version:        \t407b315e7edcbc9eb73527b904172e603095494e\n",
      "Substitution matrix    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Alignment format       \t4\n",
      "Format alignment output\tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov,cigar\n",
      "Translation table      \t1\n",
      "Gap open cost          \taa:11,nucl:5\n",
      "Gap extension cost     \taa:1,nucl:2\n",
      "Database output        \tfalse\n",
      "Preload mode           \t0\n",
      "Search type            \t0\n",
      "Threads                \t6\n",
      "Compressed             \t0\n",
      "Verbosity              \t3\n",
      "\n",
      "[=================================================================] 1 0s 0ms\n",
      "Time for merging to 1pga_A_6hka_A_resultDB.tab: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "\n",
    "# lets run MMseqs on the atom sequences\n",
    "\n",
    "WORK_DIR=\"/local/sieg/mmseqs_workdir/\"\n",
    "\n",
    "mmseqs createdb \"${WORK_DIR}/1pga_A.fasta\" \"${WORK_DIR}/1pga_A_DB\"\n",
    "mmseqs createdb \"${WORK_DIR}/6hka_A.fasta\" \"${WORK_DIR}/6hka_A_DB\"\n",
    "\n",
    "mkdir -p \"${WORK_DIR}/results\"\n",
    "\n",
    "mmseqs search \"${WORK_DIR}/1pga_A_DB\" \"${WORK_DIR}/6hka_A_DB\" \"${WORK_DIR}/results/1pga_A_6hka_A_resultDB\" \"${WORK_DIR}/tmp\" -a --max-seqs 1000000 #--cov-mode 0 --min-seq-id 0 -s 5.7\n",
    " \n",
    "mmseqs convertalis --format-mode 4 --format-output \"query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov,cigar\"\\\n",
    "                   \"${WORK_DIR}/1pga_A_DB\" \"${WORK_DIR}/6hka_A_DB\" \"${WORK_DIR}/results/1pga_A_6hka_A_resultDB\" \"${WORK_DIR}/results/1pga_A_6hka_A_resultDB.tab\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ab727049",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "query\ttarget\tfident\talnlen\tmismatch\tgapopen\tqstart\tqend\ttstart\ttend\tevalue\tbits\tqcov\ttcov\tcigar\r\n"
     ]
    }
   ],
   "source": [
    "!cat /local/sieg/mmseqs_workdir/results/1pga_A_6hka_A_resultDB.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f6cd99",
   "metadata": {},
   "source": [
    "As we can observe, when we use the atom sequence with MMseqs (the same sequence that is used by MicroMiner and TMalign) no hits are reported for the chain pari 1pga_A 6hka_A"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
